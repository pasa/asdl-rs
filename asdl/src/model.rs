#[derive(Debug)]
pub struct Asdl {
    pub types: Vec<Type>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn get_type_by_name(&self, name: &str) -> Option<&Type> {
        self.types.iter().find(|t| t.id() == name)
    }
}

#[derive(Debug)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl Type {
    fn id(&self) -> String {
        match self {
            Type::SumType(sty) => sty.id.clone(),
            Type::ProdType(pty) => pty.id.clone(),
        }
    }
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub arity: Arity,
}

#[derive(Debug)]
pub enum Arity {
    Optional,
    Required,
    Repeated,
}
