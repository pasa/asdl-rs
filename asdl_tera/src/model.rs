use std::collections::HashMap;
use serde::Serialize;
use asdl;
use super::Result;

#[derive(Serialize, Debug)]
pub struct Asdl {
    pub types: HashMap<String, Type>,
    pub prod_types: Vec<String>,
    pub sum_types: Vec<String>,
    pub comments: Vec<String>,
}

impl Asdl {
    pub fn parse(asdl: &str) -> Result<Asdl> {
        Ok(Asdl::new(asdl::Asdl::parse(asdl)?))
    }
}

#[derive(Serialize, Debug)]
#[serde(untagged)]
pub enum Type {
    SumType(SumType),
    ProdType(ProdType),
}

impl From<SumType> for Type {
    fn from(n: SumType) -> Type {
        Type::SumType(n)
    }
}
impl From<ProdType> for Type {
    fn from(n: ProdType) -> Type {
        Type::ProdType(n)
    }
}

#[derive(Serialize, Debug)]
pub struct SumType {
    pub id: String,
    pub constructors: Vec<Constructor>,
    pub attributes: Vec<Field>,
    pub is_prod_type: bool, //always false
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Constructor {
    pub id: String,
    pub fields: Vec<Field>,
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct ProdType {
    pub id: String,
    pub fields: Vec<Field>,
    pub is_prod_type: bool, //always true
    pub comments: Vec<String>,
}

#[derive(Serialize, Debug)]
pub struct Field {
    pub id: String,
    pub type_id: String,
    pub is_required: bool,
    pub is_optional: bool,
    pub is_repeated: bool,
}
