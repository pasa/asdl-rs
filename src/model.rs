use std::fmt::Write;

use rustc_hash::FxHashMap;
use heck::MixedCase;
use linked_hash_set::LinkedHashSet;
use serde::Serialize;

use crate::parser;

#[derive(Serialize, Debug)]
pub struct Asdl {
    sum_types: Vec<SumType>,
    prod_types: Vec<ProdType>,
}

#[derive(Serialize, Debug)]
pub(crate) struct SumType {
    id: String,
    constructors: Vec<Constructor>,
    attributes: Vec<Field>,
}

#[derive(Serialize, Debug)]
pub(crate) struct Constructor {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct ProdType {
    id: String,
    fields: Vec<Field>,
}

#[derive(Serialize, Debug, Clone, Hash, Eq, PartialEq)]
pub(crate) struct Field {
    id: String,
    type_id: String,
    is_single: bool,
    is_option: bool,
    is_sequence: bool,
}

impl Field {
    fn single(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: true, is_option: false, is_sequence: false }
    }

    fn option(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: true, is_sequence: false }
    }

    fn sequence(id: String, type_id: String) -> Field {
        Field { id, type_id, is_single: false, is_option: false, is_sequence: true }
    }
}

impl Asdl {
    pub(crate) fn new(root: &parser::Root) -> Self {
        let mut res = Asdl { prod_types: vec![], sum_types: vec![] };
        for d in root.types() {
            match d.kind() {
                parser::TypeKind::SumType(t) => {
                    res.sum_types.push(sum_type(t));
                }
                parser::TypeKind::ProdType(t) => {
                    res.prod_types.push(prod_type(t));
                }
            }
        }
        res
    }
}

fn sum_type(node: &parser::SumType) -> SumType {
    let id = node.type_id().text().to_string();
    let constructors = node.constructors().map(constructor).collect();
    let attributes = if let Some(attrs) = node.attrs() {
        let mut names = FieldNames::default();
        attrs.fields().map(|f| field(f, &mut names)).collect()
    } else {
        vec![]
    };
    SumType { id, constructors, attributes }
}

fn constructor(node: &parser::Constr) -> Constructor {
    let id = node.id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    Constructor { id, fields }
}

fn field(node: &parser::Field, names: &mut FieldNames) -> Field {
    match node.kind() {
        parser::FieldKind::Single(f) => Field::single(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Opt(f) => Field::option(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
        parser::FieldKind::Sequence(f) => Field::sequence(
            names.get_or_generate(f.id(), f.type_id()),
            f.type_id().text().to_string(),
        ),
    }
}

fn prod_type(node: &parser::ProdType) -> ProdType {
    let id = node.type_id().text().to_string();
    let mut names = FieldNames::default();
    let fields: Vec<Field> = node.fields().map(|f| field(f, &mut names)).collect();
    ProdType { id, fields }
}

#[derive(Default)]
struct FieldNames {
    names_indexes: FxHashMap<String, u32>,
}

impl FieldNames {
    fn get_or_generate(&mut self, id: Option<&parser::Id>, type_id: &parser::TypeId) -> String {
        match id {
            Option::Some(id) => id.text().to_string(),
            Option::None => {
                let type_id = type_id.text();
                let index = self.names_indexes.entry(type_id.to_string()).or_insert(0);
                let res = if *index == 0 {
                    type_id.to_string()
                } else {
                    let mut buf = String::new();
                    write!(&mut buf, "{}{}", type_id, index).unwrap();
                    buf
                };
                *index += 1;
                res
            }
        }
    }
}
