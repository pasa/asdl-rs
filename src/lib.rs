mod model;
mod parser;

use tera::*;
use crate::model::*;
use failure::format_err;
use heck::{CamelCase, ShoutySnakeCase, SnakeCase, MixedCase};

pub fn generate(asdl: &str, template: &str) -> String {
    let root: &parser::Root = &parser::parse(&asdl);
    let model = Asdl::from(root);
    let mut tera = Tera::default();
    tera.register_filter("camel", |arg, _| Ok(arg.as_str().unwrap().to_camel_case().into()));
    tera.register_filter("snake", |arg, _| Ok(arg.as_str().unwrap().to_snake_case().into()));
    tera.register_filter("mixed", |arg, _| Ok(arg.as_str().unwrap().to_mixed_case().into()));
    tera.register_filter("SCREAM", |arg, _| {
        Ok(arg.as_str().unwrap().to_shouty_snake_case().into())
    });
    tera.add_raw_template("_src", &template)
        .map_err(|e| format_err!("template parsing error: {:?}", e))
        .unwrap();
    tera.render("_src", &model)
        .map_err(|e| format_err!("template rendering error: {:?}", e))
        .unwrap()
}

#[cfg(test)]
mod tests {
    use crate::*;
    use insta::assert_snapshot_matches;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn simple_successful_test() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
            noFileds = One | Two | Tree
            prodType = (noFileds? f, stm s1)
            ";
        let root: &parser::Root = &parser::parse(&asdl);
        assert_snapshot_matches!("simple_successful_test_syntax", root.debug_dump());
        let model = Asdl::from(root);
        assert_debug_snapshot_matches!("simple_successful_test_model", model)
    }
}
