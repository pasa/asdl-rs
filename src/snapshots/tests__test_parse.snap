---
created: "2019-06-04T15:16:51.597004596Z"
creator: insta@0.8.1
source: src/lib.rs
expression: res.debug_dump()
---
SyntaxKind(21)@[0; 169)
  SyntaxKind(8)@[0; 13)
  SyntaxKind(22)@[13; 86)
    SyntaxKind(6)@[13; 16)
    SyntaxKind(8)@[16; 17)
    SyntaxKind(9)@[17; 18)
    SyntaxKind(8)@[18; 19)
    SyntaxKind(23)@[19; 86)
      SyntaxKind(25)@[19; 43)
        SyntaxKind(7)@[19; 27)
        SyntaxKind(26)@[27; 43)
          SyntaxKind(0)@[27; 28)
          SyntaxKind(27)@[28; 34)
            SyntaxKind(6)@[28; 31)
            SyntaxKind(8)@[31; 32)
            SyntaxKind(28)@[32; 34)
          SyntaxKind(5)@[34; 35)
          SyntaxKind(8)@[35; 36)
          SyntaxKind(27)@[36; 42)
            SyntaxKind(6)@[36; 39)
            SyntaxKind(8)@[39; 40)
            SyntaxKind(28)@[40; 42)
          SyntaxKind(1)@[42; 43)
      SyntaxKind(8)@[43; 60)
      SyntaxKind(2)@[60; 61)
      SyntaxKind(8)@[61; 62)
      SyntaxKind(25)@[62; 73)
        SyntaxKind(7)@[62; 68)
        SyntaxKind(26)@[68; 73)
          SyntaxKind(0)@[68; 69)
          SyntaxKind(27)@[69; 72)
            SyntaxKind(6)@[69; 72)
          SyntaxKind(1)@[72; 73)
      SyntaxKind(8)@[73; 86)
  SyntaxKind(22)@[86; 125)
    SyntaxKind(6)@[86; 94)
    SyntaxKind(8)@[94; 95)
    SyntaxKind(9)@[95; 96)
    SyntaxKind(8)@[96; 97)
    SyntaxKind(23)@[97; 125)
      SyntaxKind(25)@[97; 101)
        SyntaxKind(7)@[97; 100)
        SyntaxKind(8)@[100; 101)
      SyntaxKind(2)@[101; 102)
      SyntaxKind(8)@[102; 103)
      SyntaxKind(25)@[103; 106)
        SyntaxKind(7)@[103; 106)
      SyntaxKind(2)@[106; 107)
      SyntaxKind(8)@[107; 108)
      SyntaxKind(25)@[108; 125)
        SyntaxKind(7)@[108; 112)
        SyntaxKind(8)@[112; 125)
  SyntaxKind(22)@[125; 156)
    SyntaxKind(6)@[125; 133)
    SyntaxKind(8)@[133; 134)
    SyntaxKind(9)@[134; 135)
    SyntaxKind(8)@[135; 136)
    SyntaxKind(24)@[136; 156)
      SyntaxKind(26)@[136; 156)
        SyntaxKind(0)@[136; 137)
        SyntaxKind(27)@[137; 147)
          SyntaxKind(6)@[137; 145)
          SyntaxKind(8)@[145; 146)
          SyntaxKind(28)@[146; 147)
        SyntaxKind(5)@[147; 148)
        SyntaxKind(8)@[148; 149)
        SyntaxKind(27)@[149; 155)
          SyntaxKind(6)@[149; 152)
          SyntaxKind(8)@[152; 153)
          SyntaxKind(28)@[153; 155)
        SyntaxKind(1)@[155; 156)
  SyntaxKind(8)@[156; 169)

